crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -193,19 +193,19 @@ where
             self.canonical_in_memory_state.get_canonical_block_number()
         });
 
-        for num in start..=end {
-            if let Some(block_state) = self.canonical_in_memory_state.state_by_number(num) {
-                // TODO: there might be an update between loop iterations, we
-                // need to handle that situation.
-                headers.push(block_state.block().block().header.header().clone());
-            } else {
-                let mut db_headers = self.database.headers_range(num..=end)?;
+        // Perform range lookup in the in-memory state
+        match self.canonical_in_memory_state.range_lookup(start, end) {
+            RangeLookupResult::None => Ok(Vec::new()),
+            RangeLookupResult::Partial(mut in_memory_headers) => {
+                // Retrieve remaining headers from the database
+                let mut db_headers =
+                    self.database.headers_range(start + in_memory_headers.len() as u64..=end)?;
```

Comment:
this is backwards because the database has blocks lower than in memory, so we need to update the range accordingly based on the lowest and highest block in the partial range

Reply:
So,
```
self.database.headers_range(highest - 1 as u64..=end)?;
```
?